init: remove the need for a dummy device pointer in SYS_INIT functions

gmarull opened the pull request:
**NOTE**: This is a breaking change! All `SYS_INIT` functions will need
to be converted to the new signature. See the script offered in https://github.com/zephyrproject-rtos/zephyr/pull/51217/commits/c5c637c8408696eafde9c27b7c84b2dd653dc41e.

The init infrastructure, found in `init.h`, is currently used by:

- `SYS_INIT`: to call functions before `main`
- `DEVICE_*`: to initialize devices

They are all sorted according to an initialization level + a priority.
`SYS_INIT` calls are really orthogonal to devices, however, the required
function signature requires a `const struct device *dev` as a first
argument. The only reason for that is because the same init machinery is
used by devices, so we have something like:

```c
struct init_entry {
	int (*init)(const struct device *dev);
	/* only set by DEVICE_*, otherwise NULL */
	const struct device *dev;
}
```

As a result, we end up with such weird/ugly pattern:

```c
static int my_init(const struct device *dev)
{
	/* always NULL! add ARG_UNUSED to avoid compiler warning */
	ARG_UNUSED(dev);
	...
}
```

This is really a result of poor internals isolation. This patch proposes
a to make init entries more flexible so that they can accept sytem
initialization calls like this:

```c
static int my_init(void)
{
	...
}
```

This is achieved using a union:

```c
union init_function {
	/* for SYS_INIT, used when init_entry.dev == NULL */
	int (*sys)(void);
	/* for DEVICE*, used when init_entry.dev != NULL */
	int (*dev)(const struct device *dev);
};

struct init_entry {
	/* stores init function (either for SYS_INIT or DEVICE*) */
	union init_function init_fn;
	/* stores device pointer for DEVICE*, NULL for SYS_INIT. Allows
	 * to know which union entry to call.
	 */
	const struct device *dev;
}
```

This solution **does not increase ROM usage**, and allows to offer clean
public APIs for both SYS_INIT and DEVICE*. Note that however, init
machinery keeps a coupling with devices.

**NOTE**: This is a breaking change! All `SYS_INIT` functions will need
to be converted to the new signature. See the script offered in https://github.com/zephyrproject-rtos/zephyr/pull/51217/commits/c5c637c8408696eafde9c27b7c84b2dd653dc41e.

henrikbrixandersen commented:
Looks good!

galak commented on include/zephyr/init.h:
```diff
@@ -92,20 +87,40 @@ struct init_entry {
+/**
+ * @brief Define wrapper function for SYS_INIT() and SYS_INIT_NAMED().
+ *
+ * @param init_id Init entry unique identifier.
```
Should we use name instead of init_id? (since we use name for the param everywhere else)

galak commented:
LGTM!

JordanYates commented on include/zephyr/device.h:
```diff
@@ -962,6 +962,35 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
+ * @param init_fn Init function.
+ */
+#define Z_DEVICE_INIT_WRAPPER_FN_DEFINE(dev_id, init_fn)                       \
+	static inline int Z_DEVICE_INIT_WRAPPER_FN_NAME(dev_id)(void *data)    \
```
Isn't this going to result in an instance of this function for every device in the build?

carlescufi replied:
It  is indeed, this should be moved to `device.c` as a singleton instead.

JordanYates replied:
That can't be done without storing the `init_fn` in `struct device`, the singleton would have no way to retrieve the function otherwise

teburd commented:
This is nice

JordanYates commented:
I'm not against this solution, but I'm not the biggest fan of function code that sits in a macro definition.
An alternative would be each "user" of init (SYS_INIT, device) defines an init function which all instances of that user call. The `data` parameter would then be user specific.

For SYS_INIT:
```
static int sys_init_run(void *data)
{
    int(*fn)(void) = data;
    return fn();
}

struct init_entry some_sys_init = {.init = sys_init_run, .data = some_module_fn};
```

For devices:
```
static int device_init_run(void *data)
{
    const struct device *dev = data;
    int ret;

    ret = dev->init(dev);
    ...
    return ret;
}

struct init_entry some_dev_init = {.init = device_init_run, .data = dev_ptr};
```
The device solution requires storing the init function pointer in `struct device`, but this should be strictly less ROM than the proposed function wrapper. As a bonus it should allow progress on 5 from #22941, as the init function can be obtained from code.

gmarull commented:
> I'm not against this solution, but I'm not the biggest fan of function code that sits in a macro definition. An alternative would be each "user" of init (SYS_INIT, device) defines an init function which all instances of that user call. The `data` parameter would then be user specific.
> 
> For SYS_INIT:
> 
> ```
> static int sys_init_run(void *data)
> {
>     int(*fn)(void) = data;
>     return fn();
> }
> 
> struct init_entry some_sys_init = {.init = sys_init_run, .data = some_module_fn};
> ```
> 

Afaik, conversion from pointer to a function to void * is undefined behavior. (ref MISRA 11.1).

> For devices:
> 
> ```
> static int device_init_run(void *data)
> {
>     const struct device *dev = data;
>     int ret;
> 
>     ret = dev->init(dev);
>     ...
>     return ret;
> }
> 
> struct init_entry some_dev_init = {.init = device_init_run, .data = dev_ptr};
> ```
> 
> The device solution requires storing the init function pointer in `struct device`, but this should be strictly less ROM than the proposed function wrapper. As a bonus it should allow progress on 5 from #22941, as the init function can be obtained from code.

I'll try this one and see.

gmarull commented:
@JordanYates updated proposal (devices part)

gmarull commented:
> > Afaik, conversion from pointer to a function to void * is undefined behavior. (ref MISRA 11.1).
> 
> Easy to work around if we want to avoid the jump instruction per `SYS_INIT`
> 
> ```
> union init_data {
>     void *data;
>     void (*fn)(void);
> };
> 
> struct init_entry {
>     int (*init)(union init_data *data);
>     union init_data data;
> };
> 
> void sys_init_run(union init_data *data) {
>     data->fn();
> }
> 
> struct init_entry some_sys_init = {.init = sys_init_run, .data = {.fn = some_module_fn}};
> ```

good point, union will do the trick. Let me implement this.

gmarull commented:
> > Afaik, conversion from pointer to a function to void * is undefined behavior. (ref MISRA 11.1).
> 
> Easy to work around if we want to avoid the jump instruction per `SYS_INIT`
> 
> ```
> union init_data {
>     void *data;
>     void (*fn)(void);
> };
> 
> struct init_entry {
>     int (*init)(union init_data *data);
>     union init_data data;
> };
> 
> void sys_init_run(union init_data *data) {
>     data->fn();
> }
> 
> struct init_entry some_sys_init = {.init = sys_init_run, .data = {.fn = some_module_fn}};
> ```

After a quick try https://github.com/teslabs/zephyr/commit/6db182d4a16a69f5891b5834e84713d6d5203a6c, I'm not convinced about this. It also seems to be using more ROM.

gmarull committed:
scripts: utils: add sys_init migration script

Utility script to migrate SYS_INIT() calls from the old signature:

```c
int (*init_fn)(const struct device *dev)
```
to

```c
int (*init_fn)(void)
```

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

andyross reviewed (commented):
Looks great to me

galak reviewed (commented):
Are we now increasing the footprint since every device needs to now store an init pointer?

galak commented:
I can understand wanting to change the 'struct device *' to a 'void *' for the SYS_INIT callback, but I don't think we should support different callback signatures like this.  The prototype for SYS_INIT function should be consistent, and if that means we need `ARG_UNUSED(arg)` than that is ok.

gmarull commented:
> I can understand wanting to change the 'struct device *' to a 'void *' for the SYS_INIT callback, but I don't think we should support different callback signatures like this. The prototype for SYS_INIT function should be consistent, and if that means we need `ARG_UNUSED(arg)` than that is ok.

My quick measurements show little flash usage increase with current approach, e.g.
```
west build -b nrf5340dk_nrf5340_cpuapp samples/bluetooth/direction_finding_central

# main
Memory region         Used Size  Region Size  %age Used
           FLASH:      112304 B         1 MB     10.71%
             RAM:       26024 B       448 KB      5.67%
        IDT_LIST:          0 GB         2 KB      0.00%

# this PR (+14bytes FLASH)
Memory region         Used Size  Region Size  %age Used
           FLASH:      112320 B         1 MB     10.71%
             RAM:       26024 B       448 KB      5.67%
        IDT_LIST:          0 GB         2 KB      0.00%
```
Numbers are worse when using a non-common call for devices (ie an inline per device). Note that we're using per-device inlines in the case of I2C/CAN, so this PR is not really introducing any new technique.

The whole point of this PR is to remove redundant/weird argument from SYS_INIT. Here, a wrapper is used to achieve the goal, which at the same time decouples init logic from devices (see device init logic is no longer in init.c). It's the same idea as Python decorators, just more clunky.

If we're not worried about decoupling, there's another solution that allows to solve the SYS_INIT _problem_:

```c
struct init_entry {
	union {
		/* used when init_entry.dev == NULL (ie, SYS_INIT) */
		int (*sys_init)(void);
		/* used when init_entry.dev  != NULL (ie, devices) */
		int (*device_init)(const struct device *dev);
	} init_fn;
	const struct device *dev;
};
```

but IMO is not as clean as the current proposal.

gmarull commented:
> API meeting:
> 
> * @bjarki-trackunit argues that it will be necessary to add support additional signature in the future

If that's the case, current approach scales to any number of signatures. Note that, however, because this is really an internal implementation detail we can now take the union solution and modify it later. SYS_INIT/DEVICE API users will not notice.

> * @galak agrees, and advocates for using a standard, neutral signature like `int (*init)(void *arg);` instead of hardcoding the two signatures (one for sys_init, one for devices)
> * In general there is an argument about whether devices should use the init infrastructure
> * @galak says that sys_init should have a single, consistent API/infrastructure, so either `void *` or leave it alone

`void *` would lead to this pattern:

```c
int my_device_init(void *data)
{
	const struct device *dev = data;
}

int my_sys_init(void *data)
{
	ARG_UNUSED(data);
}
```

So, IMO worse than what we have, and would break both SYS_INIT and DEVICE API users. I'd like to note that the "init entry" infrastructure is a kind of implementation detail on how sys inits/devices are initialized in the end. We just propagated its internals to SYS_INIT, leading to the ugly `int my_sys_init(const struct device *dev)` pattern. This PR just makes "init entry" more flexible so that we can design clean public APIs for both, SYS_INIT and DEVICE. IMO, the sys_init call signature we have today is just a consequence of poor internals isolation.

> * @carlescufi argues that we might need two separate lists instead

We should see the resulting ROM usage. init entries for sys_inits now come with an unnecessary data/device pointer (wasted ROM), but I guess we'd have some other overhead from extra logic?

> * @yperess suggests using `int (*init)(void);`  `void *get_init_pointer()` from a global context (initialization is single-threaded)

I assume this would consume more ROM? I'd need to see a more detailed example.

> * Users are warming up to [this proposal](https://github.com/zephyrproject-rtos/zephyr/pull/51217#issuecomment-1282514717) but there is the issue of compatibility: the change in signatures for functions passed to `SYS_INIT` would require changes to out-of-tree users. @nashif is worried about this

Historically, we have not been really consistent when it comes to device macros (similar case). The last breaking change I can remember was the PM resources field. Before, device macros instantiated it for every device, now, device is expected to allocate resources and pass a pointer instead. We kept the `DEVICE_*` macros, breaking their arguments. I think it is hard to justify changing macro names just because we introduce a breaking change. Sometimes this is easy, sometimes not. We could probably come with `SYS_INIT_DEFINE` as a new macro and deprecate `SYS_INIT`, but, what will be the next name if the macro changes in the future? If, for example, we remove priorities from `DEVICE_DT_DEFINE`, will we rename the macro as well?  Really? This PR also comes with a script to automate migration, so you can update your project in literally a few seconds. Other similar breaking changes did not have such _luxury_, and they still got in.

gmarull committed:
init: remove the need for a dummy device pointer in SYS_INIT functions

The init infrastructure, found in `init.h`, is currently used by:

- `SYS_INIT`: to call functions before `main`
- `DEVICE_*`: to initialize devices

They are all sorted according to an initialization level + a priority.
`SYS_INIT` calls are really orthogonal to devices, however, the required
function signature requires a `const struct device *dev` as a first
argument. The only reason for that is because the same init machinery is
used by devices, so we have something like:

```c
struct init_entry {
	int (*init)(const struct device *dev);
	/* only set by DEVICE_*, otherwise NULL */
	const struct device *dev;
}
```

As a result, we end up with such weird/ugly pattern:

```c
static int my_init(const struct device *dev)
{
	/* always NULL! add ARG_UNUSED to avoid compiler warning */
	ARG_UNUSED(dev);
	...
}
```

This is really a result of poor internals isolation. This patch proposes
a to make init entries more flexible so that they can accept sytem
initialization calls like this:

```c
static int my_init(void)
{
	...
}
```

This is achieved using a union:

```c
union init_function {
	/* for SYS_INIT, used when init_entry.dev == NULL */
	int (*sys)(void);
	/* for DEVICE*, used when init_entry.dev != NULL */
	int (*dev)(const struct device *dev);
};

struct init_entry {
	/* stores init function (either for SYS_INIT or DEVICE*)
	union init_function init_fn;
	/* stores device pointer for DEVICE*, NULL for SYS_INIT. Allows
	 * to know which union entry to call.
	 */
	const struct device *dev;
}
```

This solution **does not increase ROM usage**, and allows to offer clean
public APIs for both SYS_INIT and DEVICE*. Note that however, init
machinery keeps a coupling with devices.

**NOTE**: This is a breaking change! All `SYS_INIT` functions will need
to be converted to the new signature. See the script offered in the
following commit.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

init: convert SYS_INIT functions to the new signature

Conversion scripted using scripts/utils/migrate_sys_init.py.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

manifest: update projects for SYS_INIT changes

Update modules with updated SYS_INIT calls:

- hal_ti
- lvgl
- sof
- TraceRecorderSource

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

tests: devicetree: devices: adjust test

Adjust test according to the recently introduced SYS_INIT
infrastructure.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

tests: kernel: threads: adjust SYS_INIT call

Adjust to the new signature: int (*init_fn)(void);

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
Updated PR with the union approach. ROM delta is now zero.

de-nordic reviewed (dismissed):
subsys/mgmt/mcumgr looks OK

nashif reviewed (changes requested):
We need a deprecation strategy.

gmarull commented:
> We need a deprecation strategy.

Can you please suggest one?

carlescufi commented:
> > We need a deprecation strategy.
> 
> Can you please suggest one?

There's only one way to do this: find a new name and deprecate the old one.

Options I can think of for the new name:

- `K_INIT`
- `Z_INIT` (we need to change the `z_` policy then)
- `SYS_INIT_CALL`
- `SYS_INIT_CALLBACK`
- `INIT_CALLBACK`

gmarull commented:
> > > We need a deprecation strategy.
> > 
> > 
> > Can you please suggest one?
> 
> There's only one way to do this: find a new name and deprecate the old one.
> 
> Options I can think of for the new name:
> 
> * `K_INIT`
> * `Z_INIT` (we need to change the `z_` policy then)
> * `SYS_INIT_CALL`
> * `SYS_INIT_CALLBACK`
> * `INIT_CALLBACK`

As I said earlier, we haven't followed this rule in some occasions, like when DEVICE* PM flag was changed to be a reference. IMO, it is better to break that entering the name lottery for this sort of changes. This is in the end improving what was an initial questionable design. What will happen if init levels are removed? Yet another name? I thought everybody agreed that it was okay to break things for the "new" device model (and sys_init is related, we have many pseudo-devices using it).

nashif commented:
> As I said earlier, we haven't followed this rule in some occasions, like when DEVICE* PM flag was changed to be a reference.

That does not make it right. And we need to stop doing this. PM for a long time was considered unstable or experimental. Zephyr is not anyone's playground, we have users and products  and every change like this withou deprecation and consideration is just an obstacle we are adding and preventing our users from smoothly migrating and moving between releases. 
We already have many ugly/weird APIs in many places, but those are not broken, are they? We should avoid changing code and breaking users just for the sake of change...

> What will happen if init levels are removed? Yet another name?

Given that this change only deals with "weird/ugly patterns" and does not provide any functional change or any improvement, maybe it should be tabled untill we are ready to redesign the whole init process and just do it once. Right now, this is just going to cause noise and breakage EVERYWHERE without any added value and you want to do all of this without deprecation on top of that.

> I thought everybody agreed that it was okay to break things for the "new" device model (and sys_init is related, we have many pseudo-devices using it).

I do not know about that and I am not sure who "everybody" is here.

carlescufi commented:
>We already have many ugly/weird APIs in many places, but those are not broken, are they? We should avoid changing code and breaking users just for the sake of change...

The problem is the fine line between "for the sake of change" and "for more logical APIs that make sense and don't force the user to work around them". Having `ARG_UNUSED` in every single `SYS_INIT` callback because it takes a `struct device*` makes no sense, we all agree. This is a good attempt at introducing common sense in that particular API, which is a good step in the right direction.

>Given that this change only deals with "weird/ugly patterns" and does not provide any functional change or any improvement, maybe it should be tabled untill we are ready to redesign the whole init process and just do it once

I disagree. We discussed this in the Architecture WG last week and the consensus seemed to be to move forward with this change as long as we could address the backwards compatibility. Iterative development is how progress happens, I doubt we will ever be in a position to "redesign it once" in the case of the init process, it's not a single driver API, it affects almost everything. So I instead propose we advance iteratively, step by step, and this PR is one step forward. The only way we have to achieve that in a backwards-compatible manner, given the limitations of the C programming language, is to rename the macro and deprecate the old one. If we can agree to "version" those then it'd be quite easy (`SYS_INIT2`, `SYS_INIT3`, etc). Otherwise we can get creative with naming.

henrikbrixandersen commented:
> The only way we have to achieve that in a backwards-compatible manner, given the limitations of the C programming language, is to rename the macro and deprecate the old one. If we can agree to "version" those then it'd be quite easy (`SYS_INIT2`, `SYS_INIT3`, etc). Otherwise we can get creative with naming.

Another option for deprecation could be to introduce a Kconfig for selecting the deprecated behaviour.

nashif commented:
> This is ignoring the fact that this PR provides a mechanism to migrate automatically. And btw, talking about playground, what about #49458? Is then right to self-merge core changes without proper reviewing because a particular company needs them?

This is not nice and IMO out of line. The above PR was up for review since August and was merged just recently, saying it went in without proper review is flat wrong. You might disagree with certain aspects of it or the naming used, but the one single patch adding a new init level does provide needed functionality and is not limited to "one company". Furthermore, it does not break any out of tree users and it does not change any APIs or changes hundreds of files. 

Maybe the init patch should have been split into its own PR and reviewed seperately, but this was certainly not an attempt to workaround the system and get what "one company" needs into the codebase.

nashif commented:
> I disagree. We discussed this in the Architecture WG last week and the consensus seemed to be to move forward with this change as long as we could address the backwards compatibility.

I disagree there was consensus. The need for this change and what value it brings was questioned by multiple people, including myself. 
Iterative development is fine as long as you do not have collateral damage every step of the way. I specifically asked the question if this change is part of something bigger and the answer was NO. Above, the deprecation is bering resisted with the argument that we might have to redesign the init levels and that we will have to change the API again. So, why deal with API change and deprecation if we know the init level are going to be on the chopping block? I am aware of multiple discussions around this, why dont we plan ahead and do the API change and deprecation just once?

gmarull commented:
> As I see it we have these options going forward:
> 
> 1. Start "versioning" certain APIs, introduce `SYS_INIT2`, deprecate `SYS_INIT`
> 2. Come up with a new name (eg. `K_INIT`), deprecate `SYS_INIT`
> 3. Use the [stable API change procedure](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes)

Since my vote is to change the API, I've opened https://github.com/zephyrproject-rtos/zephyr/issues/51609.

gmarull commented:
> > I disagree. We discussed this in the Architecture WG last week and the consensus seemed to be to move forward with this change as long as we could address the backwards compatibility.
> 
> I disagree there was consensus. The need for this change and what value it brings was questioned by multiple people, including myself. Iterative development is fine as long as you do not have collateral damage every step of the way. I specifically asked the question if this change is part of something bigger and the answer was NO. Above, the deprecation is bering resisted with the argument that we might have to redesign the init levels and that we will have to change the API again. So, why deal with API change and deprecation if we know the init level are going to be on the chopping block? I am aware of multiple discussions around this, why dont we plan ahead and do the API change and deprecation just once?

I think the answer is that this change is worth on its own, and it's not a change incompatible with other changes being discussed. And I have doubts about removal of init levels/priorities in the short/midterm. It may not even happen for SYS_INITs, who knows.

carlescufi commented:
Arch WG meeting:

- Deadline for introducing this change has been agreed to be the 3.4.0 release (before its feature freeze)

tbursztyka commented:
@carlescufi @gmarull :

What about deprecating SYS_INIT in favor of SYSTEM_INIT? 
It's more in-line with what we do in device at least: we have DEVICE_INIT not DEV_INIT. 

Same semantic as before, just a clearer name. (no need to mangle with Z_ or adding more words that would complicate the meaning imo).

stephanosio commented:
> What about deprecating SYS_INIT in favor of SYSTEM_INIT?

Do you then propose to rename everything else with the `SYS_` prefix as such? (e.g. `SYS_MUTEX_DEFINE` -> `SYSTEM_MUTEX_DEFINE`, `SYS_SEM_DEFINE` -> `SYSTEM_SEM_DEFINE` or even `SYSTEM_SEMAPHORE_DEFINE`)? Because, if not, that would introduce yet another consistency problem.

carlescufi commented:
> > What about deprecating SYS_INIT in favor of SYSTEM_INIT?
> 
> Do you then propose to rename everything else with the `SYS_` prefix as such? (e.g. `SYS_MUTEX_DEFINE` -> `SYSTEM_MUTEX_DEFINE`, `SYS_SEM_DEFINE` -> `SYSTEM_SEM_DEFINE` or even `SYSTEM_SEMAPHORE_DEFINE`)? Because, if not, that would introduce yet another consistency problem.

Yes, that is one issue. And then there's the other one, which is that even if we deprecate `SYS_INIT` and introduce a `SYSTEM_INIT`, if we do it too early we risk having to deprecate `SYSTEM_INIT` as well if we need a change in the macro. Hence the timeline.

jfischer-no commented:
> Yes, that is one issue. And then there's the other one, which is that even if we deprecate `SYS_INIT` and introduce a `SYSTEM_INIT`, if we do it too early we risk having to deprecate `SYSTEM_INIT` as well if we need a change in the macro. Hence the timeline.

I suggest to introduce `SYSTEM_INIT` as experimental feature.

gmarull commented:
> > Yes, that is one issue. And then there's the other one, which is that even if we deprecate `SYS_INIT` and introduce a `SYSTEM_INIT`, if we do it too early we risk having to deprecate `SYSTEM_INIT` as well if we need a change in the macro. Hence the timeline.
> 
> I suggest to introduce `SYSTEM_INIT` as experimental feature.

I think this discussion is really nonsense. If we think this change is good enough, then let's merge it. In this case, migration is trivial and automated using a script. IMO, the level of bureaucracy in Zephyr is reaching a level where even small changes are a pain. And there are many ways to break users, see for example https://github.com/zephyrproject-rtos/zephyr/pull/52194. It changes DT properties, so any application using such sensor will need to forcibly update DT, or build will fail. We should be more pragmatic, deprecate when possible/easy, and for changes that fix or improve previous designs allow breaking and say "Sorry, our previous design wasn't great. We're fixing it now, here's how to migrate".

tbursztyka reviewed (dismissed):
If we can keep SYS_INIT name, then fine by me.

(that said, if changing name is necessary for deprecation, why SYS_ to SYSTEM would need to be applied on all macros starting with this prefix? If we start applying such rule, it will really hard to change any name afterwads)

nordicjm commented:
OK with change, just concerned that by having it done in 2 parts, this will cause problems for people doing a `git bisect`

gmarull commented:
> OK with change, just concerned that by having it done in 2 parts, this will cause problems for people doing a `git bisect`

we can squash before merging, I did it like this to ease review.

firscity reviewed (dismissed):
Xen-related parts LGTM

soburi reviewed (dismissed):
LGTM gd32v related files

drivers/interrupt_controller/intc_nuclei_eclic.c
soc/riscv/riscv-privilege/gd32vf103/soc.c

carlocaione reviewed (dismissed):
LGTM for the files I own.

cfriedt reviewed (dismissed):
Overall, LGTM - thanks @gmarull for paying off the interest on this tech debt. `hal_ti` changes look good as well.

nashif commented:
@carlescufi  it is worth it to put this on Arch WG agenda as a reminder for those who did not follow this last year.

gmarull committed:
doc: release-notes: inform about SYS_INIT changes

Inform about SYS_INIT callback signature change. Also mention the
provided utility script to automatically migrate existing projects.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
last push: updated manifest with SHAs, rebased and added release notes.

gmarull commented:
last push: fixed some remaining issues with sof, should now be ready

henrikbrixandersen reviewed (changes requested):
> OK with change, just concerned that by having it done in 2 parts, this will cause problems for people doing a `git bisect`

The commits needs squashing before merging this to avoid `git bisect` issues.

gmarull commented:
> > OK with change, just concerned that by having it done in 2 parts, this will cause problems for people doing a `git bisect`
> 
> The commits needs squashing before merging this to avoid `git bisect` issues.

yup, this was actually mentioned/proposed. I've squashed to preserve bisectability now that everything is ready.

fabiobaltieri reviewed (approved):
Drop RFC from the PR title?

marc-hb commented:
Since this commit SOF compilation with `-Werror` fails as below.

For exact reproduction steps stay tuned.

```
In file included from /zep_workspace/sof/zephyr/include/rtos/init.h:9, 
448 from /zep_workspace/sof/src/init/init.c:13: 
449/zep_workspace/zephyr/include/zephyr/init.h:156:44: error: initialization of 'int (*)(void)' from incompatible pointer type 'int (*)(const struct device *)' [-Werror=incompatible-pointer-types] 
450 156 | .init_fn = {.sys = (init_fn_)}, \ 
451 | ^ 
452/zep_workspace/zephyr/include/zephyr/init.h:137:9: note: in expansion of macro 'SYS_INIT_NAMED' 
453 137 | SYS_INIT_NAMED(init_fn, init_fn, level, prio) 
454 | ^~~~~~~~~~~~~~ 
455/zep_workspace/sof/src/init/init.c:350:1: note: in expansion of macro 'SYS_INIT' 
456 350 | SYS_INIT(sof_init, POST_KERNEL, 99); 
457 | ^~~~~~~~ 
458/zep_workspace/zephyr/include/zephyr/init.h:156:44: note: (near initialization for '__init_sof_init.init_fn.sys') 
459 156 | .init_fn = {.sys = (init_fn_)}, \ 
460 | ^ 
461/zep_workspace/zephyr/include/zephyr/init.h:137:9: note: in expansion of macro 'SYS_INIT_NAMED' 
462 137 | SYS_INIT_NAMED(init_fn, init_fn, level, prio) 
463 | ^~~~~~~~~~~~~~ 
464/zep_workspace/sof/src/init/init.c:350:1: note: in expansion of macro 'SYS_INIT' 
465 350 | SYS_INIT(sof_init, POST_KERNEL, 99); 
466 | ^~~~~~~~ 
467cc1: all warnings being treated as errors
```

marc-hb commented:
> For exact reproduction steps stay tuned.

OK that was really easy, nothing special required:

```
west build -b intel_adsp_cavs25 ../modules/audio/sof/app -- -DEXTRA_CFLAGS=-Werror

zephyr/include/zephyr/init.h:156:44: error: initialization of 'int (*)(void)' from incompatible pointer type 'int (*)(const struct device *)' [-Werror=incompatible-pointer-types]
  156 |                         .init_fn = {.sys = (init_fn_)},          
```

Yet https://github.com/zephyrproject-rtos/zephyr/actions/runs/4678432008/jobs/8288559016 did not care.

Twister used to compile with `-Werror`, I know because a long time ago I added `'-Wa,--fatal-warnings` to it in 8a603da6cde63a380. I don't when this was lost but it was a pretty big loss IMHO.

Twister also failed to catch this regression: https://github.com/zephyrproject-rtos/zephyr/pull/56701
 
cc: @nashif, @galak

marc-hb commented:
Just spotted this sorry "NOTE: This is a breaking change! All SYS_INIT functions will need..."

The 267 comments put me off.

Still, dropping `-Werror` is not the solution to "support" backwards incompatible changes in CI.

nashif commented:
> Since this commit SOF compilation with `-Werror` fails as below.

there is a PR submitted to SOF to deal with this.

nashif commented:
> Twister also failed to catch this regression: #56701 (demoed again in #56791)

catch it where? File a bug please instead of commenting in merged PRs

dbaluta commented:
@gmarull  any idea why SOF build are failing? I think this is related to your change.

https://github.com/thesofproject/sof/actions/runs/4686511690/jobs/8304685904?pr=7429

gmarull commented:
> @dbaluta for SOF the API change was submitted a long time ago by @gmarull here:
> 
> * [zephyr: update SYS_INIT calls thesofproject/sof#6480](https://github.com/thesofproject/sof/pull/6480)
> 
> Next time a similar situation arises, corresponding module changes like this one will hopefully be present in the module branch used by Zephyr first (and required by Zephyr CI). Among others this should preserve bisectability on both sides (as long as `west update` is invoked at every step)

As a breaking change, the update needs to happen in multiple steps here. First in Zephyr fork, then upstream, then sync Zephyr fork. I don't see any better way if we allow Zephyr glue code in modules that have an upstream counterpart and that are tested in Zephyr CI.
